File: lib/ExpediaBookings/src/main/java/com/expedia/bookings/services/LXServices.java
package com.expedia.bookings.services;

import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.joda.time.LocalDate;

import com.expedia.bookings.data.Money;
import com.expedia.bookings.data.cars.ApiError;
import com.expedia.bookings.data.cars.BaseApiResponse;
import com.expedia.bookings.data.lx.ActivityDetailsResponse;
import com.expedia.bookings.data.lx.AvailabilityInfo;
import com.expedia.bookings.data.lx.LXActivity;
import com.expedia.bookings.data.lx.LXCategoryMetadata;
import com.expedia.bookings.data.lx.LXCategorySortOrder;
import com.expedia.bookings.data.lx.LXCheckoutParams;
import com.expedia.bookings.data.lx.LXCheckoutResponse;
import com.expedia.bookings.data.lx.LXCreateTripParams;
import com.expedia.bookings.data.lx.LXCreateTripResponse;
import com.expedia.bookings.data.lx.LXRedemptionType;
import com.expedia.bookings.data.lx.LXSearchParams;
import com.expedia.bookings.data.lx.LXSearchResponse;
import com.expedia.bookings.data.lx.LXSortFilterMetadata;
import com.expedia.bookings.data.lx.LXSortType;
import com.expedia.bookings.data.lx.Offer;
import com.expedia.bookings.data.lx.Ticket;
import com.expedia.bookings.utils.CollectionUtils;
import com.expedia.bookings.utils.DateUtils;
import com.expedia.bookings.utils.LXUtils;
import com.expedia.bookings.utils.Strings;
import com.squareup.okhttp.OkHttpClient;

import retrofit.RequestInterceptor;
import retrofit.RestAdapter;
import retrofit.client.OkClient;
import rx.Observable;
import rx.Observer;
import rx.Scheduler;
import rx.Subscription;
import rx.functions.Action1;
import rx.functions.Func2;

public class LXServices {

	private static LXSearchResponse cachedLXSearchResponse = new LXSearchResponse();
	LXApi lxApi;

	private OkHttpClient client;
	private Scheduler observeOn;
	private Scheduler subscribeOn;

	public LXServices(String endPoint, OkHttpClient client, RequestInterceptor requestInterceptor, Scheduler observeOn, Scheduler subscribeOn, RestAdapter.LogLevel logLevel) {
		this.client = client;
		this.observeOn = observeOn;
		this.subscribeOn = subscribeOn;

		RestAdapter adapter = new RestAdapter.Builder()
			.setEndpoint(endPoint)
			.setRequestInterceptor(requestInterceptor)
			.setLogLevel(logLevel)
			.setClient(new OkClient(this.client))
			.build();

		lxApi = adapter.create(LXApi.class);
	}

	public Subscription lxCategorySearch(LXSearchParams searchParams, Observer<LXSearchResponse> observer) {
		return lxApi
			.searchLXActivities(searchParams.location, searchParams.toServerStartDate(), searchParams.toServerEndDate())
			.doOnNext(HANDLE_SEARCH_ERROR)
			.doOnNext(ACTIVITIES_MONEY_TITLE)
			.doOnNext(PUT_POPULARITY_COUNTER_FOR_SORT)
			.doOnNext(CACHE_SEARCH_RESPONSE)
			.doOnNext(PUT_ACTIVITIES_IN_CATEGORY)
			.doOnNext(PUT_CATEGORY_KEY_IN_CATEGORY_METADATA)
			.doOnNext(PUT_CATEGORY_SORT_IN_CATEGORY_METADATA)
			.subscribeOn(subscribeOn)
			.observeOn(observeOn)
			.subscribe(observer);
	}

	public Subscription lxSearch(LXSearchParams searchParams, Observer<LXSearchResponse> observer) {
		return lxSearch(searchParams)
			.subscribe(observer);
	}

	public Observable<LXSearchResponse> lxSearch(LXSearchParams searchParams) {
		return lxApi
			.searchLXActivities(searchParams.location, searchParams.toServerStartDate(), searchParams.toServerEndDate())
			.observeOn(this.observeOn)
			.subscribeOn(this.subscribeOn)
			.doOnNext(HANDLE_SEARCH_ERROR)
			.doOnNext(ACTIVITIES_MONEY_TITLE)
			.doOnNext(CACHE_SEARCH_RESPONSE);
	}

	private static final Action1<LXSearchResponse> HANDLE_SEARCH_ERROR = new Action1<LXSearchResponse>() {
		@Override
		public void call(LXSearchResponse lxSearchResponse) {
			if (lxSearchResponse.searchFailure) {
				ApiError apiError = new ApiError(ApiError.Code.LX_SEARCH_NO_RESULTS);
				apiError.regionId = lxSearchResponse.regionId;
				ApiError.ErrorInfo errorInfo = new ApiError.ErrorInfo();
				errorInfo.cause = "No results from api.";
				apiError.errorInfo = errorInfo;
				throw apiError;
			}
		}
	};

	private static final Action1<LXSearchResponse> PUT_POPULARITY_COUNTER_FOR_SORT = new Action1<LXSearchResponse>() {
		@Override
		public void call(LXSearchResponse lxSearchResponse) {
			int popularityForClientSort = 0;
			for (LXActivity activity : lxSearchResponse.activities) {
				activity.popularityForClientSort = popularityForClientSort++;
			}
		}
	};

	private static final Action1<LXSearchResponse> PUT_CATEGORY_KEY_IN_CATEGORY_METADATA = new Action1<LXSearchResponse>() {
		@Override
		public void call(LXSearchResponse lxSearchResponse) {
			for (Map.Entry<String, LXCategoryMetadata> filterCategory : lxSearchResponse.filterCategories.entrySet()) {
				String categoryKeyEN = filterCategory.getKey();
				LXCategoryMetadata categoryValue = filterCategory.getValue();
				categoryValue.categoryKeyEN = categoryKeyEN;
			}
		}
	};

	private static final Action1<LXSearchResponse> PUT_CATEGORY_SORT_IN_CATEGORY_METADATA = new Action1<LXSearchResponse>() {
		@Override
		public void call(LXSearchResponse lxSearchResponse) {
			for (Map.Entry<String, LXCategoryMetadata> filterCategory : lxSearchResponse.filterCategories.entrySet()) {
				String categoryKeyEN = filterCategory.getKey();
				LXCategoryMetadata categoryValue = filterCategory.getValue();
				for (LXCategorySortOrder sortOrder : LXCategorySortOrder.values()) {
					if (LXUtils.whitelistAlphanumericFromCategoryKey(categoryKeyEN)
						.equalsIgnoreCase(sortOrder.toString())) {
						categoryValue.sortOrder = sortOrder;
						break;
					}
				}
			}
		}
	};

	private static final Action1<LXSearchResponse> PUT_ACTIVITIES_IN_CATEGORY = new Action1<LXSearchResponse>() {
		@Override
		public void call(LXSearchResponse lxSearchResponse) {
			for (Map.Entry<String, LXCategoryMetadata> filterCategory : lxSearchResponse.filterCategories.entrySet()) {
				String categoryKeyEN = filterCategory.getKey();
				LXCategoryMetadata categoryValue = filterCategory.getValue();
				for (LXActivity activity: lxSearchResponse.activities) {
					if (CollectionUtils.isNotEmpty(activity.categories) && activity.categories.contains(categoryKeyEN)) {
						categoryValue.activities.add(activity);
					}
				}
			}
		}
	};

	public Subscription lxDetails(String activityId, String location, LocalDate startDate, LocalDate endDate,
		Observer<ActivityDetailsResponse> observer) {
		return lxApi
			.activityDetails(activityId, location, DateUtils.convertToLXDate(startDate),
				DateUtils.convertToLXDate(endDate))
			.observeOn(this.observeOn)
			.subscribeOn(this.subscribeOn)
			.doOnNext(HANDLE_ACTIVITY_DETAILS_ERROR)
			.doOnNext(ACCEPT_ONLY_KNOWN_TICKET_TYPES)
			.doOnNext(DETAILS_TITLE_AND_TICKET_MONEY)
			.subscribe(observer);
	}

	private static final Action1<ActivityDetailsResponse> HANDLE_ACTIVITY_DETAILS_ERROR = new Action1<ActivityDetailsResponse>() {
		@Override
		public void call(ActivityDetailsResponse response) {
			if (response == null || response.offersDetail == null || response.offersDetail.offers == null) {
				throw new ApiError(ApiError.Code.LX_DETAILS_FETCH_ERROR);
			}
		}
	};

	// Add money in offer tickets for easier handling and remove &quot; from activity and offer title.
	private static final Action1<ActivityDetailsResponse> DETAILS_TITLE_AND_TICKET_MONEY = new Action1<ActivityDetailsResponse>() {
		@Override
		public void call(ActivityDetailsResponse response) {
			String bags = response.bags;
			String passengers = response.passengers;
			boolean isGroundTransport = response.isGroundTransport;
			LXRedemptionType redemptionType = response.redemptionType;

			response.title = Strings.escapeQuotes(response.title);

			for (Offer offer : response.offersDetail.offers) {
				offer.title = Strings.escapeQuotes(offer.title);
				offer.bags = bags;
				offer.passengers = passengers;
				offer.isGroundTransport = isGroundTransport;
				offer.redemptionType = redemptionType;
				for (AvailabilityInfo availabilityInfo : offer.availabilityInfo) {
					for (Ticket ticket : availabilityInfo.tickets) {
						ticket.money = new Money(ticket.amount, response.currencyCode);
					}
				}
			}
		}
	};

	private static final Action1<ActivityDetailsResponse> ACCEPT_ONLY_KNOWN_TICKET_TYPES = new Action1<ActivityDetailsResponse>() {
		@Override
		public void call(ActivityDetailsResponse response) {
			Iterator<Offer> offerIterator = response.offersDetail.offers.iterator();
			//Iterate over offers
			while (offerIterator.hasNext()) {
				Offer offer = offerIterator.next();
				Iterator<AvailabilityInfo> availabilityInfoIterator = offer.availabilityInfo.iterator();
				//Iterate over Offers's Availability Infos
				while (availabilityInfoIterator.hasNext()) {
					AvailabilityInfo availabilityInfo = availabilityInfoIterator.next();
					Iterator<Ticket> ticketIterator = availabilityInfo.tickets.iterator();
					//Iterate over Offers's Availability Info's Tickets
					while (ticketIterator.hasNext()) {
						if (ticketIterator.next().code == null) {
							//Remove Unknown Ticket Code
							ticketIterator.remove();
						}
					}

					//In case no tickets are known, rid off the availability info
					if (availabilityInfo.tickets.size() == 0) {
						availabilityInfoIterator.remove();
					}
				}

				//In case no availability info exists, rid off the offer
				if (offer.availabilityInfo.size() == 0) {
					offerIterator.remove();
				}
			}
		}
	};

	public Subscription createTrip(LXCreateTripParams createTripParams, Money originalPrice, Observer<LXCreateTripResponse> observer) {
		return lxApi.
			createTrip(createTripParams)
			.doOnNext(HANDLE_ERRORS)
			.doOnNext(new SearchOriginalPriceInjector(originalPrice))
			.observeOn(this.observeOn)
			.subscribeOn(this.subscribeOn)
			.subscribe(observer);
	}

	public Subscription lxCheckout(LXCheckoutParams checkoutParams , Observer<LXCheckoutResponse> observer) {
		Money originalPrice = new Money(checkoutParams.expectedTotalFare, checkoutParams.expectedFareCurrencyCode);
		return lxApi.
			checkout(checkoutParams.toQueryMap())
			.doOnNext(HANDLE_ERRORS)
			.doOnNext(new CreateTripOriginalPriceInjector(originalPrice))
			.observeOn(this.observeOn)
			.subscribeOn(this.subscribeOn)
			.subscribe(observer);
	}

	private static final Action1<BaseApiResponse> HANDLE_ERRORS = new Action1<BaseApiResponse>() {
		@Override
		public void call(BaseApiResponse response) {
			if (response.hasErrors() && !response.hasPriceChange()) {
				throw response.getFirstError();
			}
		}
	};

	private class SearchOriginalPriceInjector implements Action1<LXCreateTripResponse> {
		private Money originalPrice;

		public SearchOriginalPriceInjector(Money originalPrice) {
			this.originalPrice = originalPrice;
		}

		@Override
		public void call(LXCreateTripResponse lxCreateTripResponse) {
			//Set Original Price in case there was a Price Change
			if (lxCreateTripResponse.hasPriceChange()) {
				lxCreateTripResponse.originalPrice = originalPrice;
			}
		}
	}

	private class CreateTripOriginalPriceInjector implements Action1<LXCheckoutResponse> {
		Money originalPrice;

		public CreateTripOriginalPriceInjector(Money originalPrice) {
			this.originalPrice = originalPrice;
		}

		@Override
		public void call(LXCheckoutResponse lxCheckoutResponse) {
			if (lxCheckoutResponse.hasPriceChange()) {
				lxCheckoutResponse.originalPrice = originalPrice;
			}
		}
	}

	// Add money in tickets for easier handling and remove &quot; from activity title.
	private static final Action1<LXSearchResponse> ACTIVITIES_MONEY_TITLE = new Action1<LXSearchResponse>() {
		@Override
		public void call(LXSearchResponse response) {
			String currencyCode = response.currencyCode;
			for (LXActivity activity : response.activities) {
				activity.price = new Money(activity.fromPriceValue, currencyCode);
				activity.originalPrice = new Money(activity.fromOriginalPriceValue, currencyCode);
				activity.title = Strings.escapeQuotes(activity.title);
			}
		}
	};

	private class CombineSearchResponseAndSortFilterStreams implements Func2<LXSearchResponse, LXSortFilterMetadata, LXSearchResponse> {

		@Override
		public LXSearchResponse call(LXSearchResponse lxSearchResponse, LXSortFilterMetadata lxSortFilterMetadata) {

			if (lxSortFilterMetadata.lxCategoryMetadataMap == null) {
				// No filters Applied.
				lxSearchResponse.activities.clear();
				lxSearchResponse.activities.addAll(lxSearchResponse.unFilteredActivities);
				for (Map.Entry<String, LXCategoryMetadata> filterCategory : lxSearchResponse.filterCategories
					.entrySet()) {
					LXCategoryMetadata lxCategoryMetadata = filterCategory.getValue();
					lxCategoryMetadata.checked = false;
				}
			}
			else {
				lxSearchResponse.activities = applySortFilter(lxSearchResponse.unFilteredActivities, lxSearchResponse,
					lxSortFilterMetadata);
			}
			return lxSearchResponse;
		}
	}

	private class SortCategorySearchResponse implements Func2<LXCategoryMetadata, LXSortType, LXCategoryMetadata> {

		@Override
		public LXCategoryMetadata call(LXCategoryMetadata lxCategoryMetadata, LXSortType lxSortType) {
			if (lxSortType == LXSortType.PRICE) {
				Collections.sort(lxCategoryMetadata.activities, new Comparator<LXActivity>() {
					@Override
					public int compare(LXActivity lhs, LXActivity rhs) {
						Money leftMoney = lhs.price;
						Money rightMoney = rhs.price;
						return leftMoney.compareTo(rightMoney);
					}
				});
			}
			else if (lxSortType == LXSortType.POPULARITY) {
				Collections.sort(lxCategoryMetadata.activities, new Comparator<LXActivity>() {
					@Override
					public int compare(LXActivity lhs, LXActivity rhs) {
						return (lhs.popularityForClientSort < rhs.popularityForClientSort) ? -1 :
							((lhs.popularityForClientSort == rhs.popularityForClientSort) ? 0 : 1);
					}
				});
			}
			return lxCategoryMetadata;
		}
	}

	public Subscription lxCategorySort(LXCategoryMetadata category, LXSortType lxSortType, Observer<LXCategoryMetadata> categorySortObserver) {

		return (Observable.combineLatest(Observable.just(category), Observable.just(lxSortType),
			new SortCategorySearchResponse())
			.subscribeOn(this.subscribeOn)
				.observeOn(this.observeOn)
				.subscribe(categorySortObserver));
	}

	public Subscription lxSearchSortFilter(LXSearchParams lxSearchParams, LXSortFilterMetadata lxSortFilterMetadata,
		Observer<LXSearchResponse> searchResultFilterObserver) {
		Observable<LXSearchResponse> lxSearchResponseObservable =
			lxSearchParams == null ? Observable.just(cachedLXSearchResponse) : lxSearch(lxSearchParams);

		return (lxSortFilterMetadata != null ?
			Observable.combineLatest(lxSearchResponseObservable, Observable.just(lxSortFilterMetadata),
				new CombineSearchResponseAndSortFilterStreams()) :
			lxSearch(lxSearchParams))
			.doOnNext(new IsFromCachedResponseInjector(lxSearchParams == null))
			.subscribeOn(this.subscribeOn)
			.observeOn(this.observeOn)
			.subscribe(searchResultFilterObserver);
	}

	private class IsFromCachedResponseInjector implements Action1<LXSearchResponse> {
		private boolean isFromCachedResponse;

		IsFromCachedResponseInjector(boolean isFromCachedResponse) {
			this.isFromCachedResponse = isFromCachedResponse;
		}

		@Override
		public void call(LXSearchResponse lxSearchResponse) {
			lxSearchResponse.isFromCachedResponse = isFromCachedResponse;
		}
	}
	public List<LXActivity> applySortFilter(List<LXActivity> unfilteredActivities, LXSearchResponse lxSearchResponse, LXSortFilterMetadata lxSortFilterMetadata) {

		Set<LXActivity> filteredSet = new LinkedHashSet<>();
		for (int i = 0; i < unfilteredActivities.size(); i++) {
			for (Map.Entry<String, LXCategoryMetadata> filterCategory : lxSortFilterMetadata.lxCategoryMetadataMap.entrySet()) {
				LXCategoryMetadata lxCategoryMetadata = filterCategory.getValue();
				String lxCategoryMetadataKey = filterCategory.getKey();
				if (lxCategoryMetadata.checked) {
					if (unfilteredActivities.get(i).categories.contains(lxCategoryMetadataKey)) {
						filteredSet.add(unfilteredActivities.get(i));
					}
				}
			}
		}

		lxSearchResponse.activities.clear();

		// Filtering.
		if (filteredSet.size() != 0) {
			lxSearchResponse.activities.addAll(filteredSet);
		}
		else {
			lxSearchResponse.activities.addAll(unfilteredActivities);
		}

		// Sorting.
		if (lxSortFilterMetadata.sort == LXSortType.PRICE) {
			Collections.sort(lxSearchResponse.activities, new Comparator<LXActivity>() {
				@Override
				public int compare(LXActivity lhs, LXActivity rhs) {
					Money leftMoney = lhs.price;
					Money rightMoney = rhs.price;
					return leftMoney.compareTo(rightMoney);
				}
			});
		}
		return lxSearchResponse.activities;
	}

	private static final Action1<LXSearchResponse> CACHE_SEARCH_RESPONSE = new Action1<LXSearchResponse>() {
		@Override
		public void call(LXSearchResponse response) {
			cachedLXSearchResponse = response;
			cachedLXSearchResponse.unFilteredActivities.clear();
			cachedLXSearchResponse.unFilteredActivities.addAll(response.activities);
		}
	};
}

File: lib/ExpediaBookings/src/main/java/com/expedia/bookings/services/LxServices.kt
package com.expedia.bookings.services

import com.expedia.bookings.extensions.ObservableOld
import com.expedia.bookings.data.ApiError
import com.expedia.bookings.data.BaseApiResponse
import com.expedia.bookings.data.Money
import com.expedia.bookings.data.lx.ActivityDetailsResponse
import com.expedia.bookings.data.lx.LXActivity
import com.expedia.bookings.data.lx.LXCheckoutParams
import com.expedia.bookings.data.lx.LXCheckoutResponse
import com.expedia.bookings.data.lx.LXCreateTripParams
import com.expedia.bookings.data.lx.LXCreateTripResponse
import com.expedia.bookings.data.lx.LXCreateTripResponseV2
import com.expedia.bookings.data.lx.LXSearchResponse
import com.expedia.bookings.data.lx.LXSortFilterMetadata
import com.expedia.bookings.data.lx.LXSortType
import com.expedia.bookings.data.lx.LxSearchParams
import com.expedia.bookings.extensions.applySortFilter
import com.expedia.bookings.extensions.subscribeObserver
import com.expedia.bookings.utils.ApiDateUtils
import com.expedia.bookings.utils.Strings
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.Scheduler
import io.reactivex.disposables.Disposable
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import org.joda.time.LocalDate
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.Collections
import java.util.Comparator
import java.util.LinkedHashSet

class LxServices(endpoint: String, okHttpClient: OkHttpClient, interceptor: Interceptor, val observeOn: Scheduler, val subscribeOn: Scheduler) {

    private var cachedLXSearchResponse = LXSearchResponse()

    val lxApi: LXApi by lazy {

        val adapter = Retrofit.Builder()
                .baseUrl(endpoint)
                .client(okHttpClient.newBuilder().addInterceptor(interceptor).build())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()

        adapter.create(LXApi::class.java)
    }

    fun lxSearch(searchParams: LxSearchParams, observer: Observer<LXSearchResponse>): Disposable {
        return lxSearch(searchParams).subscribeObserver(observer)
    }

    fun lxSearch(searchParams: LxSearchParams): Observable<LXSearchResponse> {
        return lxApi.searchLXActivities(searchParams.location, searchParams.toServerStartDate(), searchParams.toServerEndDate(), searchParams.modQualified)
                .observeOn(this.observeOn)
                .subscribeOn(this.subscribeOn)
                .doOnNext(HANDLE_SEARCH_ERROR)
                .doOnNext(ACTIVITIES_MONEY_TITLE)
                .doOnNext(CACHE_SEARCH_RESPONSE)
    }

    private val HANDLE_SEARCH_ERROR = { response: LXSearchResponse ->
        if (response.searchFailure) {
            val apiError = ApiError(ApiError.Code.LX_SEARCH_NO_RESULTS)
            apiError.regionId = response.regionId
            val errorInfo = ApiError.ErrorInfo()
            errorInfo.cause = "No results from api."
            apiError.errorInfo = errorInfo
            throw apiError
        }
    }

    fun lxDetails(activityId: String?, location: String?, startDate: LocalDate?, endDate: LocalDate?, promoPricingEnabled: Boolean,
                  promoPricingMaxDiscountPercentageEnabled: Boolean, observer: Observer<ActivityDetailsResponse>): Disposable {
        return lxApi.activityDetails(activityId, location, ApiDateUtils.localDateToyyyyMMddSafe(startDate),
                ApiDateUtils.localDateToyyyyMMddSafe(endDate), promoPricingEnabled, promoPricingMaxDiscountPercentageEnabled)
                .observeOn(this.observeOn)
                .subscribeOn(this.subscribeOn)
                .doOnNext(HANDLE_ACTIVITY_DETAILS_ERROR)
                .doOnNext(ACCEPT_ONLY_KNOWN_TICKET_TYPES)
                .doOnNext(DETAILS_TITLE_AND_TICKET_MONEY)
                .subscribeObserver(observer)
    }

    private val HANDLE_ACTIVITY_DETAILS_ERROR = { response: ActivityDetailsResponse ->
        if (response.offersDetail == null || response.offersDetail.offers == null) {
            throw ApiError(ApiError.Code.LX_DETAILS_FETCH_ERROR)
        }
    }

    // Add money in offer tickets for easier handling and remove &quot; from activity and offer title.
    private val DETAILS_TITLE_AND_TICKET_MONEY = { response: ActivityDetailsResponse ->
        val bags = response.bags
        val passengers = response.passengers
        val isGroundTransport = response.isGroundTransport
        val redemptionType = response.redemptionType

        response.title = Strings.escapeQuotes(response.title)

        for (offer in response.offersDetail.offers) {
            offer.title = Strings.escapeQuotes(offer.title)
            offer.bags = bags
            offer.passengers = passengers
            offer.isGroundTransport = isGroundTransport
            offer.redemptionType = redemptionType
            for (availabilityInfo in offer.availabilityInfo) {
                for (ticket in availabilityInfo.tickets) {
                    ticket.money = Money(ticket.amount, response.currencyCode)
                    ticket.originalPriceMoney = Money(ticket.originalAmount, response.currencyCode)
                    ticket.prices?.forEach { price ->
                        price.money = Money(price.amount, response.currencyCode)
                        price.originalPriceMoney = Money(price.originalAmount, response.currencyCode)
                    }
                }
            }
        }
    }

    private val ACCEPT_ONLY_KNOWN_TICKET_TYPES = { response: ActivityDetailsResponse ->
        val offerIterator = response.offersDetail.offers.iterator()
        //Iterate over offers
        while (offerIterator.hasNext()) {
            val offer = offerIterator.next()
            val availabilityInfoIterator = offer.availabilityInfo.iterator()
            //Iterate over Offers's Availability Infos
            while (availabilityInfoIterator.hasNext()) {
                val availabilityInfo = availabilityInfoIterator.next()
                val ticketIterator = availabilityInfo.tickets.iterator()
                //Iterate over Offers's Availability Info's Tickets
                while (ticketIterator.hasNext()) {
                    if (ticketIterator.next().code == null) {
                        //Remove Unknown Ticket Code
                        ticketIterator.remove()
                    }
                }

                //In case no tickets are known, rid off the availability info
                if (availabilityInfo.tickets.size == 0) {
                    availabilityInfoIterator.remove()
                }
            }

            //In case no availability info exists, rid off the offer
            if (offer.availabilityInfo.size == 0) {
                offerIterator.remove()
            }
        }
    }

    fun createTrip(createTripParams: LXCreateTripParams, originalPrice: Money, observer: Observer<LXCreateTripResponse>): Disposable {
        return lxApi.createTrip(createTripParams)
                .doOnNext(HANDLE_ERRORS)
                .doOnNext {
                    response: LXCreateTripResponse ->
                    if (response.hasPriceChange()) {
                        response.originalPrice = originalPrice
                    }
                }
                .observeOn(this.observeOn)
                .subscribeOn(this.subscribeOn)
                .subscribeObserver(observer)
    }

    fun createTripV2(createTripParams: LXCreateTripParams, originalPrice: Money, observer: Observer<LXCreateTripResponseV2>): Disposable {
        return lxApi.createTripV2(createTripParams)
                .doOnNext(HANDLE_ERRORS)
                .doOnNext {
                    response: LXCreateTripResponseV2 ->
                    if (response.hasPriceChange()) {
                        response.originalPrice = originalPrice
                    }
                }
                .observeOn(this.observeOn)
                .subscribeOn(this.subscribeOn)
                .subscribeObserver(observer)
    }

    fun lxCheckout(checkoutParams: LXCheckoutParams, observer: Observer<LXCheckoutResponse>): Disposable {
        val originalPrice = Money(checkoutParams.expectedTotalFare, checkoutParams.expectedFareCurrencyCode)
        return lxApi.checkout(checkoutParams.toQueryMap())
                .doOnNext(HANDLE_ERRORS)
                .doOnNext {
                    response: LXCheckoutResponse ->
                    if (response.hasPriceChange()) {
                        response.originalPrice = originalPrice
                    }
                }
                .observeOn(this.observeOn)
                .subscribeOn(this.subscribeOn)
                .subscribeObserver(observer)
    }

    private val HANDLE_ERRORS = { response: BaseApiResponse ->
        if (response.hasErrors() && !response.hasPriceChange()) {
            throw response.firstError
        }
    }

    // Add money in tickets for easier handling and remove &quot; from activity title.
    private val ACTIVITIES_MONEY_TITLE = { response: LXSearchResponse ->
        val currencyCode = response.currencyCode
        for (activity in response.activities) {
            activity.price = Money(activity.fromPriceValue, currencyCode)
            activity.originalPrice = Money(activity.fromOriginalPriceValue, currencyCode)
            activity.title = Strings.escapeQuotes(activity.title)
            activity.mipOriginalPrice = Money(activity.mipFromOriginalPriceValue, currencyCode)
            activity.mipPrice = Money(activity.mipFromPriceValue, currencyCode)
        }
    }

    private fun CombineSearchResponseAndSortFilterStreams(lxSearchResponse: LXSearchResponse, lxSortFilterMetadata: LXSortFilterMetadata): LXSearchResponse {

        if (lxSortFilterMetadata.lxCategoryMetadataMap == null) {
            // No filters Applied.
            lxSearchResponse.activities.clear()
            lxSearchResponse.activities.addAll(lxSearchResponse.unFilteredActivities)
            for (filterCategory in lxSearchResponse.filterCategories.entries) {
                val lxCategoryMetadata = filterCategory.value
                lxCategoryMetadata.checked = false
            }
        } else {
            lxSearchResponse.activities = applySortFilter(lxSearchResponse.unFilteredActivities, lxSearchResponse,
                    lxSortFilterMetadata)
        }
        return lxSearchResponse
    }

    fun lxSearchSortFilter(lxSearchParams: LxSearchParams?, lxSortFilterMetadata: LXSortFilterMetadata?,
                           searchResultFilterObserver: Observer<LXSearchResponse>, lxFilterTextSearchEnabled: Boolean): Disposable {

        val lxSearchResponseObservable = if (lxSearchParams == null)
            Observable.just<LXSearchResponse>(cachedLXSearchResponse) else lxSearch(lxSearchParams)

        return (
                if (lxSortFilterMetadata != null)
                    ObservableOld.combineLatest(lxSearchResponseObservable, Observable.just(lxSortFilterMetadata),
                            { searchResponse, sortFilterMetadata ->
                                if (lxFilterTextSearchEnabled) {
                                    searchResponse.activities = searchResponse.unFilteredActivities.applySortFilter(sortFilterMetadata)
                                    searchResponse
                                } else {
                                    CombineSearchResponseAndSortFilterStreams(searchResponse, sortFilterMetadata)
                                }
                            })
                else lxSearch(lxSearchParams!!)
                )
                .doOnNext { isFromCachedResponseInjector(lxSearchParams == null, it) }
                .subscribeOn(this.subscribeOn)
                .observeOn(this.observeOn)
                .subscribeObserver(searchResultFilterObserver)
    }

    private fun isFromCachedResponseInjector(isFromCachedResponse: Boolean, lxSearchResponse: LXSearchResponse) = {
        lxSearchResponse.isFromCachedResponse = isFromCachedResponse
    }

    private val CACHE_SEARCH_RESPONSE = { response: LXSearchResponse ->
        cachedLXSearchResponse = response
        cachedLXSearchResponse.unFilteredActivities.clear()
        cachedLXSearchResponse.unFilteredActivities.addAll(response.activities)
        cachedLXSearchResponse = response
    }

    fun applySortFilter(unfilteredActivities: List<LXActivity>, lxSearchResponse: LXSearchResponse, lxSortFilterMetadata: LXSortFilterMetadata): List<LXActivity> {

        val filteredSet = LinkedHashSet<LXActivity>()
        for (i in unfilteredActivities.indices) {
            for (filterCategory in lxSortFilterMetadata.lxCategoryMetadataMap.entries) {
                val lxCategoryMetadata = filterCategory.value
                val lxCategoryMetadataKey = filterCategory.key
                if (lxCategoryMetadata.checked) {
                    if (unfilteredActivities[i].categories.contains(lxCategoryMetadataKey)) {
                        filteredSet.add(unfilteredActivities[i])
                    }
                }
            }
        }

        lxSearchResponse.activities.clear()

        // Filtering.
        lxSearchResponse.activities.addAll(if (filteredSet.size != 0) filteredSet else unfilteredActivities)

        // Sorting.
        if (lxSortFilterMetadata.sort == LXSortType.PRICE) {
            Collections.sort<LXActivity>(lxSearchResponse.activities, object : Comparator<LXActivity> {
                override fun compare(lhs: LXActivity, rhs: LXActivity): Int {
                    val leftMoney = lhs.price
                    val rightMoney = rhs.price
                    return leftMoney.compareTo(rightMoney)
                }
            })
        }
        return lxSearchResponse.activities
    }
}
